File: 2015/10/10.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

def look_and_say(x)
  ret   = []
  count = 0
  val   = x.first

  x.each_index do |i|
    if x[i] == val
      count += 1
    else
      ret << count << val
      count, val = 1, x[i]
    end
  end
  ret << count << val

  ret
end

input = File.read("10.txt")
            .strip
            .chars

silver = 40.times { input = look_and_say input }.then { input.length }
gold   = 10.times { input = look_and_say input }.then { input.length }

puts("
 Day 10
========
✮: #{silver}
★: #{gold}")

File: 2015/10/10.clj
(require
  '[clojure.string :refer (split-lines split trim)])

(defn look-and-say [xs]
  (loop [lst xs ret []]
    (if (empty? lst)
      ret
      (let [head (first lst)
            [group tail] (split-with (partial = head) lst)]
        (recur tail (conj ret (count group) head))))))

(defn loop-and-say [xs n]
  (loop [i n ret xs]
    (if (zero? i)
      ret
      (recur (dec i) (look-and-say ret)))))

(def I (->> "10.txt" slurp trim seq (map #(Character/digit % 10))))

(def silver-seq (loop-and-say I 40))
(def silver (count silver-seq))
(def gold (count (loop-and-say silver-seq 10)))

(println "Day 10\n======")
(println (str "✮: " silver))
(println (str "★: " gold))
